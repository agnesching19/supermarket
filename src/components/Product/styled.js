import styled from 'styled-components';

export default styled.div`
  border: 1px solid lightgray;
  border-radius: 5px;
  margin: 36px 24px;
  width: 240px;
  > h4 {
    font-size: 20px;
    font-weight: normal;
    margin: 20px auto 12px;
    text-align: center;
    background-color: palegoldenrod;
  }
  > img {
    width: 100%;
    height: auto;
  }
  > .middle-div {
    height: 30px;
    > .discount {
      text-align: center;
      color: darkorange;
      background-color: lightgoldenrodyellow;
      border-radius: 15px;
      width: max-content;
      margin: 0 auto;
      padding: 5px 15px;
    }
  }
  > .unitPrice {
    text-align: right;
    padding-right: 10px;
    margin-top: 2px;
  }
  > .bottom-div {
    display: flex;
    flex-direction: row;
    justify-content: center;
    margin-bottom: 20px;
    button {
      height: 25px;
      width: 40px;
      
      border-radius: 4px;
      border: none;
      &:first-child {
        margin-right: 10px;
        background-color: steelblue;
      }
      &:nth-child(2) {
        background-color: lightsteelblue;
      }
    }
  }
`;