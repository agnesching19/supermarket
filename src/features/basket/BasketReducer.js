import { ADD_ITEM, GET_TOTAL, REMOVE_ITEM } from './actionTypes';

const initialState = {
  basketItems: [],
  total: {},
  showCheckout: false,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case ADD_ITEM:
      return { 
        ...state,
        basketItems: action.basketItems,
        showCheckout: action.showCheckout,
        total: {
          ...state.total,
          totalBeforeDiscount: action.totalBeforeDiscount,
        }
      };
    case REMOVE_ITEM:
      return {
        ...state,
        basketItems: action.basketItems,
        showCheckout: action.showCheckout,
        total: {
          ...state.total,
          totalBeforeDiscount: action.totalBeforeDiscount,
        }
      };
    case GET_TOTAL:
      return {
        ...state,
        total: action.payload,
        showCheckout: action.showCheckout,
      };
    default:
      return state;
  }
};
