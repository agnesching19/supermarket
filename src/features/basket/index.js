export {default as BasketReducer} from './BasketReducer';
export {default as useActions} from './actionCreators';
export {useBasketValue, useTotalValue, useShowCheckoutValue} from './selectors';
