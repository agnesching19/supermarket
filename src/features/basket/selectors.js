import { useSelector } from 'react-redux';

const useTotalValue = () => useSelector(state => state.checkout.total);
const useShowCheckoutValue = () => useSelector(state => state.checkout.showCheckout);
const useBasketValue = () => useSelector(state => state.checkout.basketItems);

export { useBasketValue, useTotalValue, useShowCheckoutValue };
