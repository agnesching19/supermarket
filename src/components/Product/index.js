import React from 'react';
import PropTypes from 'prop-types';
import {useActions} from '../../features/basket';
import Styled from './styled';

const Product = ({
  id,
  name,
  priceByItem,
  priceByWeight,
  unitPrice,
  weight,
  url,
  promotion,
}) => {
  const {addItem, removeItem} = useActions();

  return (
    <Styled>
      <h4>{name}</h4>
      <img src={url} alt={name} />

      <div className="middle-div">
        {promotion && (
          <p className="discount" data-testid="promotion">
            {promotion}
          </p>
        )}
      </div>

      {priceByItem && <p className="unitPrice">£{unitPrice.toFixed(2)}/each</p>}

      {priceByWeight && <p className="unitPrice">£{unitPrice.toFixed(2)}/kg</p>}

      <div className="bottom-div">
        <button
          data-testid="btn-add"
          type="button"
          onClick={() =>
            addItem(id, name, priceByItem, priceByWeight, unitPrice, promotion)
          }>
          +
        </button>

        <button
          data-testid="btn-remove"
          type="button"
          onClick={() => removeItem(id, priceByWeight, unitPrice)}>
          -
        </button>
      </div>
    </Styled>
  );
};

Product.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  priceByItem: PropTypes.bool,
  priceByWeight: PropTypes.bool,
  unitPrice: PropTypes.number,
  url: PropTypes.string,
  promotion: PropTypes.string,
  weight: PropTypes.number,
};

Product.defaultProps = {
  id: '',
  name: '',
  priceByItem: false,
  priceByWeight: false,
  unitPrice: 0,
  url: '',
  promotion: null,
  weight: null,
};

export default Product;
