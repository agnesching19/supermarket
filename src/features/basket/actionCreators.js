import {useCallback} from 'react';
import {useDispatch} from 'react-redux';
import { useBasketValue } from './selectors';
import { REMOVE_ITEM, ADD_ITEM, GET_TOTAL } from './actionTypes';

const useActions = () => {
  const dispatch = useDispatch();
  const basketItems = useBasketValue();

  const getIndex = useCallback(
    id => basketItems.map(i => i).findIndex(f => f.id === id),
    [basketItems]
  );

  const subTotal = basketItems
    .map(i => {
      if (i.priceByWeight) {
        return i.unitPrice * i.weight;
      }
      return i.unitPrice;
    })
    .reduce((ac, c) => ac + c, 0);

  /** Add item action */
  const addItem = useCallback(
    (id, name, priceByItem, priceByWeight, unitPrice, promotion) => {
      const index = getIndex(id);
      /** If basket has this item and it's priceByWeight, increase the weight by 0.2kg */
      if (priceByWeight && index > -1) {
        const item = basketItems[index];
        const newItems = basketItems.filter((b, i) => i !== index);
        dispatch({
          type: ADD_ITEM,
          totalBeforeDiscount: subTotal + unitPrice * 0.2,
          showCheckout: false,
          basketItems: [
            ...newItems,
            {
              ...item,
              weight: item.weight + 0.2,
            },
          ],
        });
      } else {
        dispatch({
          type: ADD_ITEM,
          totalBeforeDiscount: priceByWeight
            ? subTotal + unitPrice * 0.2
            : subTotal + unitPrice,
          showCheckout: false,
          basketItems: [
            ...basketItems,
            {
              id,
              name,
              priceByItem,
              priceByWeight,
              unitPrice,
              promotion,
              weight: priceByWeight ? 0.2 : null,
            },
          ],
        });
      }
    },
    [basketItems, dispatch, getIndex, subTotal]
  );

  /** Remove item action */
  const removeItem = useCallback(
    (id, priceByWeight, unitPrice) => {
      const index = getIndex(id);
      const newItems = basketItems.filter((b, i) => i !== index);
      /** If basket has this item and it's priceByWeight, decrease the weight by 0.2kg */
      if (priceByWeight && index > -1) {
        const item = basketItems[index];
        const lastByWeightItem = item.weight.toFixed(2) <= 0.2;
        
        dispatch({
          type: REMOVE_ITEM,
          totalBeforeDiscount: lastByWeightItem
            ? subTotal
            : subTotal - unitPrice * 0.2,
          showCheckout: false,
          basketItems: [
            ...newItems,
            {
              ...item,
              weight: lastByWeightItem ? 0 : item.weight - 0.2,
            },
          ],
        });
      } else {
        const item = basketItems[index];
        const hasPriceByItem = item ? subTotal - item.unitPrice : 0
        const hasPriceByWeight = item ? subTotal - item.unitPrice * 0.2 : 0;
      
        dispatch({
          type: REMOVE_ITEM,
          showCheckout: false,
          totalBeforeDiscount: priceByWeight ? hasPriceByWeight : hasPriceByItem,
          basketItems: index > -1 ? [...newItems] : [...basketItems],
        });
      }
    },
    [basketItems, dispatch, getIndex, subTotal]
  );

  /** Calculate total */
  const getTotal = useCallback(() => {
    const getSavings = () => {
      const counts = basketItems
        .map(item => item)
        .reduce(
          (acc, value) => ({
            ...acc,
            [value.id]: (acc[value.id] || 0) + 1,
          }),
          {}
        );

      /**
       * If there are more items using the same type of discount, we can use an api to store
       * and calculate the discount amount; here I just hardcode item_coke and item_beans as
       * these are what we need in this task.
       */
      const cokeSaved = counts.item_coke >= 2 ? counts.item_coke / 2 : 0;
      const itemCoke = basketItems[getIndex('item_coke')];
      const discountCoke =
        cokeSaved > 0
          ? {
              name: itemCoke && itemCoke.promotion,
              saved: itemCoke && itemCoke.unitPrice * Math.floor(cokeSaved),
            }
          : null;

      const beansSaved = counts.item_beans >= 3 ? counts.item_beans / 3 : 0;
      const itemBeans = basketItems[getIndex('item_beans')];
      const discountBeans =
        beansSaved > 0
          ? {
              name: itemBeans && itemBeans.promotion,
              saved: itemBeans && itemBeans.unitPrice * Math.floor(beansSaved),
            }
          : null;

      const discountItems = [discountCoke, discountBeans].filter(i => i);

      return discountItems;
    };

    const totalSaved = getSavings()
      .map(s => s.saved)
      .reduce((ac, c) => ac + c, 0);

    dispatch({
      type: GET_TOTAL,
      payload: {
        discount: getSavings(),
        totalBeforeDiscount: subTotal,
        totalFinal: (subTotal - totalSaved).toFixed(2),
      },
      showCheckout: true,
    });
  }, [basketItems, dispatch, getIndex, subTotal]);

  return {addItem, removeItem, getTotal};
};

export default useActions;
