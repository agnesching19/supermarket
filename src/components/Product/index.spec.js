import React from 'react';
import {Provider} from 'react-redux';
import {render, fireEvent} from '@testing-library/react';
import configureStore from 'redux-mock-store';
import {REMOVE_ITEM, ADD_ITEM} from 'features/basket/actionTypes';
import Product from './index';

const mockStore = configureStore([]);

describe('components > Product', () => {
  const store = mockStore({
    checkout: {
      basketItems: [],
    },
  });

  jest.spyOn(store, 'dispatch');

  beforeEach(() => {
    store.dispatch.mockClear();
  });

  it('renders without crashing', () => {
    const {asFragment, getByTestId} = render(
      <Product
        id="item_test"
        name="test"
        promotion="Get 3 for 2!"
        unitPrice={0.5}
        priceByItem
      />,
      {
        wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
      }
    );

    expect(asFragment()).toMatchSnapshot();

    expect(getByTestId('promotion').textContent).toBe('Get 3 for 2!');
  });

  it('dispatches an action on add item button click', () => {
    const {getByTestId} = render(
      <Product
        id="item_test"
        name="test"
        promotion="Get 3 for 2!"
        unitPrice={0.5}
        priceByItem
      />,
      {
        wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
      }
    );
    fireEvent.click(getByTestId('btn-add'));

    expect(store.dispatch).toHaveBeenCalledTimes(1);

    expect(store.dispatch).toHaveBeenCalledWith({
      type: ADD_ITEM,
      basketItems: [
        {
          id: 'item_test',
          name: 'test',
          promotion: 'Get 3 for 2!',
          unitPrice: 0.5,
          priceByItem: true,
          priceByWeight: false,
          weight: null,
        },
      ],
      showCheckout: false,
      totalBeforeDiscount: 0.5,
    });
  });

  it('dispatches an action on remove item button click', () => {
    const {getByTestId} = render(
      <Product
        id="item_test"
        name="test"
        promotion="Get 3 for 2!"
        unitPrice={0.5}
        priceByItem
      />,
      {
        wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
      }
    );
    fireEvent.click(getByTestId('btn-remove'));

    expect(store.dispatch).toHaveBeenCalledTimes(1);

    expect(store.dispatch).toHaveBeenCalledWith({
      type: REMOVE_ITEM,
      basketItems: [],
      showCheckout: false,
      totalBeforeDiscount: 0,
    });
  });
});
