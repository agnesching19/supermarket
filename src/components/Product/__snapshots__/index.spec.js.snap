// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > Product renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm hPTBmF"
  >
    <h4>
      test
    </h4>
    <img
      alt="test"
      src=""
    />
    <div
      class="middle-div"
    >
      <p
        class="discount"
        data-testid="promotion"
      >
        Get 3 for 2!
      </p>
    </div>
    <p
      class="unitPrice"
    >
      Â£0.50/each
    </p>
    <div
      class="bottom-div"
    >
      <button
        data-testid="btn-add"
        type="button"
      >
        +
      </button>
      <button
        data-testid="btn-remove"
        type="button"
      >
        -
      </button>
    </div>
  </div>
</DocumentFragment>
`;
